fair/2.0.0/tools/unifiedtools.py

```diff
@@ -0,0 +1,65 @@
+import numexpr as ne
+
+def calculate_alpha(G,G_A,T,r0,rC,rT,rA,g0,g1,iirf100_max = False):
+
+#     iirf100_val = r[...,0] + r[...,1] * (G-G_A) + r[...,2] * T + r[...,3] * G_A
+#     iirf100_val = np.abs(iirf100_val)
+#     if iirf100_max:
+#         iirf100_val = (iirf100_val>iirf100_max) * iirf100_max + iirf100_val * (iirf100_val<iirf100_max)
+#     alpha_val = g0 * np.exp(iirf100_val / g1)
+
+    iirf100_val = ne.evaluate("abs(r0 + rC * (G-G_A) + rT * T + rA * G_A)")
+    if iirf100_max:
+        iirf100_val = ne.evaluate("where(iirf100_val>iirf100_max,iirf100_max,iirf100_val)")
+    alpha_val = ne.evaluate("g0 * exp(iirf100_val / g1)")
+
+    return alpha_val
+
+def calculate_g(a,tau):
+    g1 = ne.evaluate("sum( a * tau * ( 1. - ( 1. + 100/tau ) * np.exp(-100/tau) ), axis=-1 )")
+    g0 = ne.evaluate("exp( -1 * sum( a * tau * ( 1. - exp(-100/tau) ) , axis=-1) / g1 )")
+    return g0, g1
+
+def step_concentration(emissions,a,dt,alpha,tau,R_old,PI_conc,emis2conc):
+#     decay_rate = dt/(alpha*tau)
+#     decay_factor = np.exp( -decay_rate )
+#     R_new = E * a * 1/decay_rate * ( 1. - decay_factor ) + R_old * decay_factor
+#     G_A = np.sum(R,axis=-1)
+#     C = PI_conc + emis2conc * (G_A + G_A_old) / 2
+    decay_rate = ne.evaluate("dt/(alpha*tau)")
+    decay_factor = ne.evaluate("exp(-decay_rate)")
+    R = ne.evaluate("E * a / decay_rate * ( 1. - decay_factor ) + R_old * decay_factor")
+    G_A = ne.evaluate("sum(R,axis=4")
+    C = ne.evaluate("PI_conc + emis2conc * (G_A + G_A_old) / 2")
+
+    return C,R,G_A
+
+def step_forcing(C,PI_conc,f1,f2,f3):
+    
+    # if the logarithmic/sqrt term is undefined (ie. C is zero or negative), this contributes zero to the overall forcing. An exception will appear, however.
```

Comment:
does that mean this needs to be wrapped in a try except block?

Reply:
Hmm. So I took this function from Nick's GIR with minimum alterations. That comment is about the previous implementation, which uses numpy functions rather than numexpr functions. Those don't have the conditional logic in them, so an expression is raised, and nan returned for e.g. ln(-1). Nick gets around this programmatically, but I can just delete those comments.

Reply:
Done

Reply:
> I can just delete those comments

Great

> That comment is about the previous implementation

My opinion is that comments are, mostly, a waste of time. The code should describe what it does, and if it doesn't, it probably means a new function or something needs to be introduced so the code becomes self-describing. Comments (like this one) tend to become outdated so are more trouble than they're worth

Reply:
Yeah, I think I agree